### PR TITLE
Fix collect on stateful generator

### LINKED ISSUE #35530: Stateful Iterator omitting last element when used in list comprehension, crashes on 1-element long vectors.
Strage behaviour of Stateful iterator:
Although
```julia
julia> for a in Iterators.Stateful([1,2,3])
             print(a)
         end
123
```
works as it should, 
With list comprehension it fails to iterate over last element
```julia
julia> [a for a in Iterators.Stateful([1,2,3])]
2-element Array{Int64,1}:
 1
 2
```

```julia
julia> [a for a in Iterators.Stateful([1])]
ERROR: BoundsError: attempt to access 0-element Array{Int64,1} at index [1]
```

tried on Julia 1.3.1, Julia 1.4.0 and Julia 1.4.1.
More details in https://discourse.julialang.org/t/strange-behavior-with-list-comprehensions-using-iterators-stateful/29715

### PR DESCRIPTION
**Pull request by jakobnissen**
Fixes https://github.com/JuliaLang/julia/issues/35530.

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -766,13 +766,25 @@ function collect(itr::Generator)
...
+        if y === nothing
+            return et[]
+        end
+        v1, st = y
+        return collect_to_with_first!(Vector{typeof(v1)}(undef, len), v1, itr, st)
+    elseif isa(isz, HasShape)
+        axs = axes(itr)
         y = iterate(itr)
         if y === nothing
-            return _array_for(et, itr.iter, isz)
```
Comment (vtjnash) [inline base/array.jl:772]: Also need to delete the code for `_array_for` just above?

Comment (vtjnash) [reply base/array.jl:772]: Ah, let's keep it then. It is, roughly speaking, an inlined version of this same function, for expanding sufficiently simple (syntax-wise) typed-comprehensions directly to for loops.

**Comment by jakobnissen** [comment]
The CI failure come from an an inference failure that reduces to this:
```
julia> h(x) = Array{typeof(x[]),0}
h (generic function with 1 method)

julia> only(Base.return_types(h, (Ref{Union{Nothing, Int}},)))
Type{Array{_A, 0}} where _A
```
I can try to work around it, but looking at it as a human, without understanding how the compiler works, it appears to be something that type inference ought to be able to figure out?

**Comment by jakobnissen** [comment]
Okay I decided to work around it.

**Pull request merged**